bfd/elfxx-riscv.c

```diff
@@ -1760,6 +1769,69 @@ riscv_parse_add_implicit_subsets (riscv_parse_subset_t *rps)
     riscv_parse_add_subset (rps, "zicsr",
 			    RISCV_UNKNOWN_VERSION,
 			    RISCV_UNKNOWN_VERSION, TRUE);
+  else if ((riscv_lookup_subset (rps->subset_list, "k", &subset)))
+    {
+      riscv_parse_add_subset (rps, "zkg",
+                              RISCV_UNKNOWN_VERSION,
+                              RISCV_UNKNOWN_VERSION, TRUE);
+      riscv_parse_add_subset (rps, "zkb",
+                              RISCV_UNKNOWN_VERSION,
+                              RISCV_UNKNOWN_VERSION, TRUE);
+      riscv_parse_add_subset (rps, "zkr",
+                              RISCV_UNKNOWN_VERSION,
+                              RISCV_UNKNOWN_VERSION, TRUE);
+      riscv_parse_add_subset (rps, "zkne",
+                              RISCV_UNKNOWN_VERSION,
+                              RISCV_UNKNOWN_VERSION, TRUE);
+      riscv_parse_add_subset (rps, "zknd",
+                              RISCV_UNKNOWN_VERSION,
+                              RISCV_UNKNOWN_VERSION, TRUE);
+      riscv_parse_add_subset (rps, "zknh",
+                              RISCV_UNKNOWN_VERSION,
+                              RISCV_UNKNOWN_VERSION, TRUE);
+      riscv_parse_add_subset (rps, "zksed",
+                              RISCV_UNKNOWN_VERSION,
+                              RISCV_UNKNOWN_VERSION, TRUE);
+      riscv_parse_add_subset (rps, "zksh",
+                              RISCV_UNKNOWN_VERSION,
+                              RISCV_UNKNOWN_VERSION, TRUE);
+      riscv_parse_add_subset (rps, "zknh",
```

Comment:
I'm a little confused about the imply extensions.  For g, it will be expanded to imafd, and won't be generated to the output ISA string.  For v, we won't add the implicit zvamo and zvlsseg, we use INSN_CLASS_V_OR_VAMO for instead.  It means we won't add the zvamo and zvlsseg to the ISA output ISA string when only enable v, but we can use the instructions.  Maybe it’s because v didn’t have the implicit implementation at that time.  It might be worth taking some time to clarify, I think the better rules are as follows,

* If the base extension can be divided into separate Z extensions, and be covered by them, then I prefer to think of the base extension as g.  We will add multiple implicit Z extensions for it, and then only output the Z extensions to the output ISA string.  I think b is just like this example, and k maybe also is.  I haven't see the spec of k in details, so not sure about this.

* Some of v instructions don't belong to the Zv* extensions, they belongs to base v directly, so the base v cannot be suppressed.  For those zvamo instructions, like vamoaddei8.v, we probably can change their instruction classes from INSN_CLASS_V_OR_ZVAMO to INSN_CLASS_ZVAMO, and add them as v's implicit extensions.  Therefore, we can only enable base v, and then can use zvamo and zvlsseg instructions.  Also output them to the output ISA string with the default versions.

Reply:
Yes, K is like B.  Everything is in a zk* subset, and K is just a collection of zk* subsets.  We probably should canonicalize like we do for g/imafd.

V is confused.  The zvamo chapter says "This set of instructions is included in the base "V" extension used for the Unix profile." which implies that zvamo is part of V.  But if zvamo is part of V, then there would be no reason to have a name for it.  And that leads to the question of whether zvamo part of V for embedded targets.  We have an open issue for this, riscv-v-spec 548.  Krste's proposed solution in riscv-v-spec issue 550 is to add a whole bunch of zve* extensions for embedded, in which case the zvamo name becomes useful again.  I would suggest not trying to make any generalizations based on how V works, as V is still too confused over its subextension naming scheme.

Reply:
> Yes, K is like B. Everything is in a zk* subset, and K is just a collection of zk* subsets. We probably should canonicalize like we do for g/imafd.

Thanks, this looks reasonable to me :)

> V is confused. The zvamo chapter says "This set of instructions is included in the base "V" extension used for the Unix profile." which implies that zvamo is part of V. But if zvamo is part of V, then there would be no reason to have a name for it. And that leads to the question of whether zvamo part of V for embedded targets. We have an open issue for this, riscv-v-spec 548. Krste's proposed solution in riscv-v-spec issue 550 is to add a whole bunch of zve* extensions for embedded, in which case the zvamo name becomes useful again. I would suggest not trying to make any generalizations based on how V works, as V is still too confused over its subextension naming scheme.

OK, so I think keep the current implementations until the vector spec clarify them is a wise choice.